net/src/client/tcp/actors/jsonrpc.rs

```diff
@@ -191,69 +246,73 @@ impl Message for Subscribe {
 impl Handler<Subscribe> for JsonRpcClient {
     type Result = <Subscribe as Message>::Result;
 
-    fn handle(&mut self, _msg: Subscribe, ctx: &mut Self::Context) -> Self::Result {
-        if let Some(SetSubscriber(recipient, request)) = self.subscriber.take() {
-            ctx.address()
-                .send(request.clone())
-                .map_err(|err| log::error!("Couldn't subscribe: {}", err))
-                .into_actor(self)
-                .map(|resp, act, ctx| {
-                    match resp {
-                        Ok(Value::String(id)) => {
-                            let stream = act
-                                .socket
-                                .subscribe(&id.clone().into())
-                                .map(|value| {
-                                    log::debug!(
-                                        "<< Forwarding notification from node to subscriber"
-                                    );
-                                    Notification(value)
-                                })
-                                .map_err(|err| Error::RequestFailed {
-                                    message: err.to_string(),
-                                    error_kind: err.0,
-                                });
-                            Self::add_stream(stream, ctx);
-                            act.subscription_id = Some(id);
-                            log::info!("Client subscription created");
-                        }
-                        Ok(_) => {
-                            log::error!("Unsupported subscription id. Subscription cancelled.");
-                        }
-                        Err(err) => {
-                            log::error!("Couldn't subscribe: {}", err);
-                        }
-                    };
-                })
-                .spawn(ctx);
-
-            self.subscriber = Some(SetSubscriber(recipient, request));
-        }
-    }
-}
-
-struct RetryConnect;
-
-impl Message for RetryConnect {
-    type Result = ();
-}
-
-impl Handler<RetryConnect> for JsonRpcClient {
-    type Result = <RetryConnect as Message>::Result;
-
-    fn handle(&mut self, _msg: RetryConnect, _ctx: &mut Self::Context) -> Self::Result {
-        log::info!(
-            "Client connection has failed, it will retry to re-connect in the next request."
+    fn handle(&mut self, subscribe: Subscribe, ctx: &mut Self::Context) -> Self::Result {
+        let request = subscribe.0.clone();
+        let topic = subscription_topic_from_request(&request);
+        log::debug!(
+            "Handling Subscribe message for topic {}. Request is {:?}",
+            topic,
+            request
         );
-        self.retry_connect = true;
+
+        ctx.address()
+            .send(request.clone())
```

Comment:
Use `ctx.notify()` to send messages to self:

```suggestion
        ctx.notify(request.clone())
```

Reply:
I think `ctx.notify` does `do_send` instead of `send` though, so we lose the chance to do the fancy stuff that we do here with the response.

Reply:
I mean, I thought exactly the same, put `ctx.notify` in place, realized it had a totally different API and rolled back the change so it could be refactored in a different PR.

Reply:
The actix docs mention that sending messages to self is a bad idea and that's why `ctx.notify()` exists. I don't know why, maybe it's an anti-pattern in the actor world or something.